### ISSUE TITLE
community crime example csv file

### ISSUE THREAD

**Issue by jasmainak** [glm-tools/pyglmnet#149]
I tried looking here: http://archive.ics.uci.edu/ml/machine-learning-databases/communities/ for the *.csv file in the community crime example but it's not a csv file. 

Also, looking at pandas `read_csv`, it seems that you can also provide it a url. If we do that, all users can run the example without having to download it separately ...

**Comment by pavanramkumar** [comment]
Great idea. It would be good to rewrite all real data examples with `read_csv` fetching the file directly.

**Comment by jasmainak** [comment]
Cool, do you know what is the actual url for the csv file in this case?

**Comment by pavanramkumar** [comment]
I believe that @marquesVF downloaded the `communities.data` file and converted it to a csv manually. 

Is it possible to use `wget`, fetch the dataset and use some `pandas` functions to import as data frame?

**Comment by jasmainak** [comment]
maybe it's possible. Do you know how exactly the conversion to csv was done? We could try to script it if it's easy

**Comment by pavanramkumar** [comment]
I don't know but did you try whether `pandas.read_csv()` can directly read the `communities.data` file?

**Comment by jasmainak** [comment]
yeah I tried that already but no luck

**Comment by themantalope** [comment]
I was able to get `pandas.read_csv("http://archive.ics.uci.edu/ml/machine-learning-databases/communities/communities.data")` to work on my machine, but it looks like the csv file has been included as part of the project now anyways. Should probably close this issue.

**Comment by jasmainak** [comment]
hmm ... I think you cannot expect users to go looking for the `.csv` file in the github repo. Often the user will just click on download script and run it. If you could get the URL to run, then can you make a PR to fix it? Thanks!

**Comment by pavanramkumar** [comment]
@jasmainak I think it makes sense now to have a folder called `datasets/` where the csv files live. I will be adding more csv files for other examples shortly, so it will indeed make sense to have a dataset loading API. We just need `datasets/datasets.py` with a couple of useful functions. 

Use case could be as simple as
```python
import pyglmnet.datasets as datasets
X, y = datasets.load_communities_data()
```

Under the hood, the function needs to read in with pandas and convert to numpy arrays and return.

All this is similar in flavor to `sklearn.datasets`.

**Comment by jasmainak** [comment]
@pavanramkumar dataset fetching can be quite a bit of work, so I was trying to avoid going in that direction for now. You have to decide where the data is fetched -- which would probably be configured by the user -- so you would have to save the configuration somewhere, if it has been fetched correctly so you verify the download hash etc etc. But I agree in principle that we should go in that direction in the future.

**Comment by pavanramkumar** [comment]
@jasmainak sure, that's for `version 0.2` but i think a more proximal version would involve:
- provide the csv files in `datasets/`so that they are available within the repo when the user clones `pyglmnet`.
- provide very simple `datasets.load_*()` methods for them to get started.

**Comment by jasmainak** [comment]
@pavanramkumar it's okay to add csv files if it's a couple of kilobytes. The moment, you go into MBs, you cannot remove it any more because you will basically double the size of the file in your repo when you remove it. Again, depends how much you want to limit the size of your repo. mne-python is 11 MB to clone as far as I remember ... and that includes many small files in the test folders. Right now the approach is more towards having a separate repo for the testing data and having automatic fetchers for that.

**Comment by marquesVF** [comment]
I applied some data pre-processing on the `communities.data` (removed some rows with missing values and removed some columns) before creating the csv file. I think the ideia of a `datasets` class and a `datasets` folder (containing minimal versions of csv files) is a good ideia for testing and documentation. 
As @jasmainak  said, as longs as those csv files individually stay a couple of KB.

**Comment by pavanramkumar** [comment]
@jasmainak good points! @marquesVF thanks! 

I checked that the current csv file is ~300 KB but we are going to have a couple more csv files for some examples I am working on.

What about the following as a set of tasks of `version 0.1`?
- make a separate repo called `pyglmnet-datasets`
- write some version of `wget` that fetches the data
- apply those to examples

We can add hash checking etc. in `version 0.2`.

**Comment by marquesVF** [comment]
> What about the following as a set of tasks of version 0.1?
>
> - make a separate repo called pyglmnet-datasets
> - write some version of wget that fetches the data into the user's <repo-home>/datasets/ folder
> - apply those to examples

I agree with that and I can work on this. The `wget` method would be included in the `Makefile` as a new target.
The new repo would belong to whom? (#166)

**Comment by jasmainak** [comment]
okay, if you want to share the dataset from another location, you might want to check the license conditions or check with the authors if they are okay with such resharing (if no such license exists) ...

what's the rationale behind hosting it in a separate repo vs fetching it directly from the source?

**Comment by pavanramkumar** [comment]
@marquesVF thanks for taking this on! I'll let you know as soon I've transferred the repo. You could ideally work on it this weekend.

@jasmainak good point about license to redistribute. Let me check. Advantage is preprocessing.

**Comment by pavanramkumar** [comment]
@marquesVF you can now go ahead with the dataset fetcher implementation. what about following @jasmainak 's suggestion of writing a fetcher instead of hosting third party data on a `pyglmnet-datasets` repo?

you can follow the ideas here as closely as possible: http://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_mldata.html#sklearn.datasets.fetch_mldata

so you would create a file `pyglmnet/datasets.py` which has functions to download and prepare the data. for instance, `fetch_community_crime_data()` would download the community crime dataset from the original host and then prepare it for use with `pyglmnet`.

**Comment by marquesVF** [comment]
I'm having problems with mldata.org registration in order to upload the dataset. There's any other online data repository?

**Comment by pavanramkumar** [comment]
@marquesVF why are you trying to upload the dataset? what i meant is:
- make a script called `datasets.py`
- write a class `Datasets` inside the script
- the class should have a function `fetch_community_crime_data()`
- `fetch_community_crime_data()` should use `wget` etc. to download the raw data from the original source (`https://archive.ics.uci.edu`) and then preprocess it to return `X`, `y` which can then be used by `pyglmnet`.

**Comment by marquesVF** [comment]
Oh. My bad. I misunderstood .
Alright then. I gonna work on that soon.

**Comment by pavanramkumar** [comment]
@marquesVF i created a module called `datasets.py`

https://github.com/pavanramkumar/pyglmnet/blob/50e4ca34300e6d61bd42fe7d26d684f5e01dd6aa/pyglmnet/datasets.py

(this will be merged soon)

your dataset fetcher can be a function in `datasets.py`. in my experience ,`subprocess` with `curl` worked better then `wget`.
